File: java/tst/TestSSD.java
// Supplier - Shipment - Detail example

import org.garret.perst.*;
import java.util.Iterator;
import java.io.*;

public class TestSSD extends Persistent {
    static class Supplier extends Persistent {
        String name;
        String location;
    }
    
    static class Detail extends Persistent {
        String id;
        float  weight;
    }
    
    static class Shipment extends Persistent { 
        Supplier supplier;
        Detail   detail;
        int      quantity;
        long     price;
    }

    FieldIndex supplierName;
    FieldIndex detailId;
    FieldIndex shipmentSupplier;
    FieldIndex shipmentDetail;
     
    static byte[] inputBuffer = new byte[256];

    static void skip(String prompt) {
        try { 
            System.out.print(prompt);
            System.in.read(inputBuffer);
        } catch (IOException x) {}
    }

    static String input(String prompt) {
        while (true) { 
            try { 
                System.out.print(prompt);
                int len = System.in.read(inputBuffer);
                String answer = new String(inputBuffer, 0, len).trim();
                if (answer.length() != 0) {
                    return answer;
                }
            } catch (IOException x) {}
        }
    }

    static long inputLong(String prompt) { 
        while (true) { 
            try { 
                return Long.parseLong(input(prompt), 10);
            } catch (NumberFormatException x) { 
                System.err.println("Invalid integer constant");
            }
        }
    }

    static double inputDouble(String prompt) { 
        while (true) { 
            try { 
                return Double.parseDouble(input(prompt));
            } catch (NumberFormatException x) { 
                System.err.println("Invalid floating point constant");
            }
        }
    }

    static public void main(String[] args) {    
        Storage db = StorageFactory.getInstance().createStorage();
        Supplier   supplier;
        Detail     detail;
        Shipment   shipment;
        Shipment[] shipments;
        Iterator   iterator;
        int        i;

        db.open("testssd.dbs");

        TestSSD root = (TestSSD)db.getRoot();
        if (root == null) { 
            root = new TestSSD();
            root.supplierName = db.createFieldIndex(Supplier.class, "name", true);
            root.detailId = db.createFieldIndex(Detail.class, "id", true);
            root.shipmentSupplier = db.createFieldIndex(Shipment.class, "supplier", false);
            root.shipmentDetail = db.createFieldIndex(Shipment.class, "detail", false);
            db.setRoot(root);
        }
        while (true) { 
            try { 
                switch ((int)inputLong("-------------------------------------\n" + 
                                       "Menu:\n" + 
                                       "1. Add supplier\n" + 
                                       "2. Add detail\n" + 
                                       "3. Add shipment\n" + 
                                       "4. List of suppliers\n" + 
                                       "5. List of details\n" + 
                                       "6. Suppliers of detail\n" + 
                                       "7. Details shipped by supplier\n" + 
                                       "8. Exit\n\n>>"))
                {
                  case 1:
                    supplier = new Supplier();
                    supplier.name = input("Supplier name: ");
                    supplier.location = input("Supplier location: ");
                    root.supplierName.put(supplier);
                    db.commit();
                    continue;
                  case 2:
                    detail = new Detail();
                    detail.id = input("Detail id: ");
                    detail.weight = (float)inputDouble("Detail weight: ");
                    root.detailId.put(detail);
                    db.commit();
                    continue;
                  case 3:
                    supplier = (Supplier)root.supplierName.get(new Key(input("Supplier name: ")));
                    if (supplier == null) { 
                        System.err.println("No such supplier!");
                        break;
                    }
                    detail = (Detail)root.detailId.get(new Key(input("Detail ID: ")));
                    if (detail == null) { 
                        System.err.println("No such detail!");
                        break;
                    }
                    shipment = new Shipment();
                    shipment.quantity = (int)inputLong("Shipment quantity: ");
                    shipment.price = inputLong("Shipment price: ");
                    shipment.detail = detail;
                    shipment.supplier = supplier;
                    root.shipmentSupplier.put(shipment);
                    root.shipmentDetail.put(shipment);
                    db.commit();
                    continue;
                  case 4:
                    iterator = root.supplierName.iterator();
                    while (iterator.hasNext()) { 
                        supplier = (Supplier)iterator.next();
                        System.out.println("Supplier name: " + supplier.name + ", supplier.location: " + supplier.location);
                    }
                    break;
                  case 5:
                    iterator = root.detailId.iterator();
                    while (iterator.hasNext()) { 
                        detail = (Detail)iterator.next();
                        System.out.println("Detail ID: " + detail.id + ", detail.weight: " + detail.weight);
                    }
                    break;
                  case 6:
                    detail = (Detail)root.detailId.get(new Key(input("Detail ID: ")));
                    if (detail == null) { 
                        System.err.println("No such detail!");
                        break;
                    }
                    shipments = (Shipment[])root.shipmentDetail.get(new Key(detail), new Key(detail));
                    for (i = 0; i < shipments.length; i++) { 
                        System.out.println("Suppplier name: " + shipments[i].supplier.name);
                    }
                    break;
                  case 7:
                    supplier = (Supplier)root.supplierName.get(new Key(input("Supplier name: ")));
                    if (supplier == null) { 
                        System.err.println("No such supplier!");
                        break;
                    }
                    shipments = (Shipment[])root.shipmentSupplier.get(new Key(supplier), new Key(supplier));
                    for (i = 0; i < shipments.length; i++) { 
                        System.out.println("Detail ID: " + shipments[i].detail.id);
                    }
                    break;
                  case 8:
                    db.close();
                    return;
                }
                skip("Press ENTER to continue...");
            } catch (StorageError x) { 
                System.out.println("Error: " + x.getMessage());
                skip("Press ENTER to continue...");
            }
        }
    }
}

File: csharp/examples/TestSSD/TestSSD.cs
using Volante;
using System;

public class Supplier : Persistent {
    public String name;
    public String location;
}

public class Detail : Persistent {
    public String id;
    public float  weight;
}

public class Shipment : Persistent { 
    public Supplier supplier;
    public Detail   detail;
    public int      quantity;
    public long     price;
}

public class TestSSD : Persistent {
    public IFieldIndex<string,Supplier>    supplierName;
    public IFieldIndex<string,Detail>     detailId;
    public IFieldIndex<Supplier,Shipment> shipmentSupplier;
    public IFieldIndex<Detail,Shipment>   shipmentDetail;

    static void skip(String prompt) {
        Console.Write(prompt);
        Console.ReadLine();
    }

    static String input(System.String prompt)
    {
        while (true)
        {
            Console.Write(prompt);
            String line = Console.ReadLine().Trim();
            if (line.Length != 0) 
            { 
                return line;
            }
        }
    }

    static long inputLong(String prompt) { 
        while (true) { 
            try { 
                return Int32.Parse(input(prompt));
            } catch (FormatException) { 
                Console.WriteLine("Invalid integer constant");
            }
        }
    }

    static double inputDouble(String prompt) { 
        while (true) { 
            try { 
                return Double.Parse(input(prompt));
            } catch (FormatException) { 
                Console.WriteLine("Invalid floating point constant");
            }
        }
    }

    static public void Main(String[] args) {	
        IDatabase db = DatabaseFactory.CreateDatabase();
        Supplier   supplier;
        Detail     detail;
        Shipment   shipment;
        Shipment[] shipments;
        int        i;

        db.Open("testssd.dbs");

        TestSSD root = (TestSSD)db.Root;
        if (root == null) { 
            root = new TestSSD();
            root.supplierName = db.CreateFieldIndex<string, Supplier>("name", IndexType.Unique);
            root.detailId = db.CreateFieldIndex<string, Detail>("id", IndexType.Unique);
            root.shipmentSupplier = db.CreateFieldIndex<Supplier, Shipment>("supplier", IndexType.NonUnique);
            root.shipmentDetail = db.CreateFieldIndex<Detail, Shipment>("detail", IndexType.NonUnique);
            db.Root = root;
        }
        while (true) { 
            try { 
                switch ((int)inputLong("-------------------------------------\n" + 
                                       "Menu:\n" + 
                                       "1. Add supplier\n" + 
                                       "2. Add detail\n" + 
                                       "3. Add shipment\n" + 
                                       "4. List of suppliers\n" + 
                                       "5. List of details\n" + 
                                       "6. Suppliers of detail\n" + 
                                       "7. Details shipped by supplier\n" + 
                                       "8. Exit\n\n>>"))
                {
                  case 1:
                    supplier = new Supplier();
                    supplier.name = input("Supplier name: ");
                    supplier.location = input("Supplier location: ");
                    root.supplierName.Put(supplier);
                    db.Commit();
                    continue;
                  case 2:
                    detail = new Detail();
                    detail.id = input("Detail id: ");
                    detail.weight = (float)inputDouble("Detail weight: ");
                    root.detailId.Put(detail);
                    db.Commit();
                    continue;
                  case 3:
                    supplier = root.supplierName[input("Supplier name: ")];
                    if (supplier == null) { 
                        Console.WriteLine("No such supplier!");
                        break;
                    }
                    detail = root.detailId[input("Detail ID: ")];
                    if (detail == null) { 
                        Console.WriteLine("No such detail!");
                        break;
                    }
                    shipment = new Shipment();
                    shipment.quantity = (int)inputLong("Shipment quantity: ");
                    shipment.price = inputLong("Shipment price: ");
                    shipment.detail = detail;
                    shipment.supplier = supplier;
                    root.shipmentSupplier.Put(shipment);
                    root.shipmentDetail.Put(shipment);
                    db.Commit();
                    continue;
                  case 4:
                    foreach (Supplier s in root.supplierName) { 
                        Console.WriteLine("Supplier name: " + s.name + ", supplier.location: " + s.location);
                    }
                    break;
                  case 5:
                    foreach (Detail d in root.detailId) {
                        Console.WriteLine("Detail ID: " + d.id + ", detail.weight: " + d.weight);
                    }
                    break;
                  case 6:
                    detail = root.detailId[input("Detail ID: ")];
                    if (detail == null) { 
                        Console.WriteLine("No such detail!");
                        break;
                    }
                    shipments = (Shipment[])root.shipmentDetail.Get(new Key(detail), new Key(detail));
                    for (i = 0; i < shipments.Length; i++) { 
                        Console.WriteLine("Suppplier name: " + shipments[i].supplier.name);
                    }
                    break;
                  case 7:
                    supplier = root.supplierName[input("Supplier name: ")];
                    if (supplier == null) { 
                        Console.WriteLine("No such supplier!");
                        break;
                    }
                    shipments = root.shipmentSupplier.Get(supplier, supplier);
                    for (i = 0; i < shipments.Length; i++) { 
                        Console.WriteLine("Detail ID: " + shipments[i].detail.id);
                    }
                    break;
                  case 8:
                    db.Close();
                    return;
                }
                skip("Press ENTER to continue...");
            } catch (DatabaseException x) { 
                Console.WriteLine("Error: " + x.Message);
                skip("Press ENTER to continue...");
            }
        }
    }
}

